Automattic/jetpack
# Paid Blocks Nudge: Fix Border by Injecting has-warning Style

Fixes #13007.
Fixes #13008, too :sunglasses: 

#### Changes proposed in this Pull Request:

Paid Blocks Nudge: Fix Border by Injecting has-warning Style

Prior to this PR, we used custom styling for the `jetpack-paid-block__wrapper` class, using padding and negative margins. This lead to visual glitches such as #13007:

<img width="558" alt="Screenshot 2019-07-05 at 15 56 01" src="https://user-images.githubusercontent.com/87168/60723836-6fdbad80-9f3d-11e9-9f1f-1fff88b69261.png">

As [documented there](https://github.com/Automattic/jetpack/issues/13007#issuecomment-510009903), this is because that kind of styling doesn't play well with Gutenberg's own, which is using _absolute_ positioning on `:before` selectors such as

-  `.block-editor-block-list__layout .block-editor-block-list__block .block-editor-block-list__block-edit:before` 
- `.block-editor-block-list__layout .block-editor-block-list__block.is-selected>.block-editor-block-list__block-edit:before`
- `.block-editor-block-list__layout .block-editor-block-list__block.has-warning .block-editor-block-list__block-edit:before`

So ideally, we want to piggy-back on those selectors -- the `has-warning` ones in particular, since they do precisely what we want (also on a conceptual level). However, that's not easy from within the block, since those selectors apply to the block _wrapper_. Class names for that wrapper are pretty much [hard-wired](https://github.com/WordPress/gutenberg/blob/eba2fc8ba1ffbd5997cf47b3aa7002f8a5597e9f/packages/block-editor/src/components/block-list/block.js#L392-L408), and there aren't really any props to that element that we could (or should) modify to enforce the styling we want.

But fortunately, there's a [filter](https://developer.wordpress.org/block-editor/developers/filters/block-filters/#editor-blocklistblock) to modify those wrappers :tada: 

This PR thus uses that filter to inject the `has-warning` class, fixing #13007, and allowing us to drop a bunch of custom styling.

However, the `has-warning` selector also adds styling to prevent user interactivity with the given block (see https://github.com/Automattic/jetpack/pull/13019#issuecomment-510129941 for details), so we need to override that behavior, since we want to ensure the user can continue to interact with the block. To that end, we add another class name, `is-interactive` for increase specificity, and override the styling that would otherwise prevent user interactions. 

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
Modifies the existing paid plan nudge.

#### Testing instructions:

- Switch to this branch locally, build Jetpack (`yarn build`), and start Docker (`yarn docker:up`)
- Now add the following line to any file in `docker/mu-plugins`, e.g. a newly created `0-blocks-upgrade.php` (needs to start with `<?php`): `define ( 'JETPACK_SHOW_BLOCK_UPGRADE_NUDGE', true );`
- Make sure you're connected to WP.com, and on a free plan
- Start a new post, and insert the 'Simple Payments' block
- Select the block, and enter some information into its text fields.
- Unselect the block again
- Verify that the visual glitch seen in the above screenshot is fixed.
- Verify that you can still select and edit the block.
- Finally, verify that #13008 is also fixed by this PR :tada: 

#### Proposed changelog entry for your changes:

(Covered by whatever we add for #12823)
